### ISSUE TITLE
Spread with enum index key, invalid members don't trigger errors

### ISSUE THREAD

**Issue by osdiab** [microsoft/TypeScript#32236]
**TypeScript Version:**  3.5.1

**Search Terms:**
object spread enum index missing keys 

**Code**
[Playground link has code](https://www.typescriptlang.org/play/#code/KYOwrgtgBAcg9gFwJYDMkGMCGy4gLLAAmSkUA3gFBTVQCCACvQDRU30ASA8jAKIUC+FCqFLxkaLDhAAVAJ4AHYOVbUAYpwAyGzgHUeAERY0oGgJIBpAwKFIQCYACcUmdErGoM2JLnoPgKR1BXACVgdDgHQmVqCCISCAAuWEQPSW98OMgjKAQFYCT3CS9cOUVs0EwAIwAbIiTKuDhazBBrCnCQAGcEKGAADyRu2wBzX39OgpSiqTGAvxAQsIjCAG0AXSgAXigVlWjjGljiSEnxTykCY4gAOgZmPeNcxVPU4pk86-UtXQMmKAeaBUanUoM5qp0lHtBGsANxCAD08Kgpl6fUU6B6CAAFoMcnBQZgkNU-hClJVgNU4AB3dq4bqowbIECjPwoTp4TDyeQjLYMobM2ada5+QhgVwAHj2lAONBWAGtgLIoLZkmc0rhLvE1klpTKDvLFcqQKrXlJSsBtVAGk1gC0AdR+HDjIIAHwACj2btiEDgf3krIAlFsXfsZR16SB+ggkk9gHAUFBvfjtrq9TRrhmk9k09QVv7-Ncjlqdfac4ioBnC8AfXnWVWrms-uWqUhsTkcZ0oOEILE7EQoHAwAg-tjAlAqYPqlFyYnBp0eQrZJ0SfjW1BOljJ1FnETS2nawXY5ayIm41SkvmUNcgbUovwm0iEIhMNVqkrhkgAG48qkOXDDKBPxfMAlFjKAX06fE0EjPc9UwTpCGcBCUCSAAiHFUKgctMEIYhmVRBAHEwUEkApQhO0IOBgE6EAAHIekcP8HF6VssUcWCHRoUtBDTPwEDABxjUjPoECdGh7ylfhwM7FpZAoAM4SAA)

**Expected behavior:**
Missing keys & incorrect value types when I build a nested object using the spread operator with index keys, ought to throw errors for invalid types

**Actual behavior:**
It seems like the child object can have whatever shape I want, the types aren't being checked at all

**Playground Link:** [This is same as above playground link](https://www.typescriptlang.org/play/#code/KYOwrgtgBAcg9gFwJYDMkGMCGy4gLLAAmSkUA3gFBTVQCCACvQDRU30ASA8jAKIUC+FCqFLxkaLDhAAVAJ4AHYOVbUAYpwAyGzgHUeAERY0oGgJIBpAwKFIQCYACcUmdErGoM2JLnoPgKR1BXACVgdDgHQmVqCCISCAAuWEQPSW98OMgjKAQFYCT3CS9cOUVs0EwAIwAbIiTKuDhazBBrCnCQAGcEKGAADyRu2wBzX39OgpSiqTGAvxAQsIjCAG0AXSgAXigVlWjjGljiSEnxTykCY4gAOgZmPeNcxVPU4pk86-UtXQMmKAeaBUanUoM5qp0lHtBGsANxCAD08Kgpl6fUU6B6CAAFoMcnBQZgkNU-hClJVgNU4AB3dq4bqowbIECjPwoTp4TDyeQjLYMobM2ada5+QhgVwAHj2lAONBWAGtgLIoLZkmc0rhLvE1klpTKDvLFcqQKrXlJSsBtVAGk1gC0AdR+HDjIIAHwACj2btiEDgf3krIAlFsXfsZR16SB+ggkk9gHAUFBvfjtrq9TRrhmk9k09QVv7-Ncjlqdfac4ioBnC8AfXnWVWrms-uWqUhsTkcZ0oOEILE7EQoHAwAg-tjAlAqYPqlFyYnBp0eQrZJ0SfjW1BOljJ1FnETS2nawXY5ayIm41SkvmUNcgbUovwm0iEIhMNVqkrhkgAG48qkOXDDKBPxfMAlFjKAX06fE0EjPc9UwTpCGcBCUCSAAiHFUKgctMEIYhmVRBAHEwUEkApQhO0IOBgE6EAAHIekcP8HF6VssUcWCHRoUtBDTPwEDABxjUjPoECdGh7ylfhwM7FpZAoAM4SAA)

**Related Issues:** I'm not really sure what the root limitation that's causing this is so it's hard for me to find similar issues.

**Comment by RyanCavanaugh** [comment]
```ts
enum Prop {
    Alpha,
    Beta
}

type MediumMap = {
    [key in Prop]: {
        m: string;
    };
};

declare const SomeProp: Prop;
const k1: MediumMap = {
    [Prop.Alpha]: { m: "" },
    [Prop.Beta]: { m: "" },
    // No error
    [SomeProp]: { m: 10 }
};
```

**Comment by RyanCavanaugh** [comment]
So the problem here is that there's not a good way to describe the type of `k3` in the above example - it's *technically* `{ A: string; B: number } | { A: number; B: string }` but this type algebra is combinatorially explosive in the presence of multiple computed property names, and isn't actually correct if given e.g. `{ [AB]: "", [AB]: "" }`. At least in the indirected case, it's not clear that we have any available fix.

That said, ideally we'd have some way to flag the `[AB]` property in `k1` as an error by realizing it's a contextually typed thing, keeping the literalness of `AB` around, and then making sure it's assignable to the intersection of the properties it possibly targets.

@ahejlsberg any thoughts?

**Comment by dragomirtitian** [comment]
@RyanCavanaugh  This seems to affect not just enums. If the computed property is `string` it basically gets ignored in the spread: 

```
declare let o: { field1: number };
declare let key: string;

o = { ...o, [key]: "value" };  // ok ? why ?
let newO = { ...o, [key]: "value" };  // same type as o ? but why ?

declare let oIndex: { [n: string]: number };
oIndex = { ...oIndex, [key]: "value" } // here we get an error

```
[Playground Link](https://www.typescriptlang.org/play/?ssl=5&ssc=12&pln=5&pc=36#code/CYUwxgNghgTiAEEQBd4HsBc8De8BmAliBMAIxYB2ArgLYBGIM8AvgNwBQoksCSqA1iACeWAM7IYBCgHMO7NPAC8OeADp1aADTwA2oKEBdLACIAblAhUQxlq3jwA9A-T94AfngB3ABZD37PngKEE8AeSUVdVUtXX0jeDMLKxs2eyd4USgaBGQhAAcEKFF0d3g6KlQfPzd2TnBoOEQUdABJClAADyxcHQoxCSlpeOp6Rlt5Ns6I3Ci0SZAO7T1heMTLaxZHZ29GBE8EaWaoCnhGGDQYdiA)

I really don't think the result of `{ ...o, [key]: "value" }` should be `typeof o` unless key is a specific key of `typeof o`. This is fundamentally unsound and surprising.

**Comment by osdiab** [comment]
@dragomirtitian i think technically that’s ok because the resulting object may have other keys but still satisfies the interface that the object type specifies - though it’s also the case that applying that key-value pair may cause an existing key on the object to have an incorrect value type, which is unsound

**Comment by danilowanner** [comment]
I am running into the same issue. Any update on this?

It is fairly common to update an object using code similar to this:

```ts
type Key = 'a' | 'b' | 'x'
type Obj = {[key in Key]: number}

const x = 'x' as Key

const obj: Obj = {
  a: 4,
  b: 4,
  x: 4,
}

const updatedObj: Obj = {
  ...obj,
  [x]: 'hello', // No error. No type check on the value, which is a string instead of number
}
```

The same is true for `Object.assign`:

```ts
const updatedObj: Obj = Object.assign({}, obj, { [x]: '4' }) // No error
```

It seems that `{ [x]: '4' }` gets widened to the type `{ [x: string]: string }`, but arguably something like the following would be more accurate: `{a: string} | {b: string} | {x: string}`. Even then it is currently still a valid third argument to `Object.assign`.

**This issue may be related to:**
#37103
#13948
#38663
